code/modules/food_and_drinks/kitchen_machinery/food_replicator.dm

```diff
@@ -0,0 +1,631 @@
+
+/obj/machinery/food_replicator/proc/set_broken()
+	stat |= BROKEN
+	//make a shower of sparks to communicate the wires burned out
+	var/datum/effect/system/spark_spread/s = new /datum/effect/system/spark_spread
+	s.set_up(2, 1, src)
+	s.start()
+	use_power = 0	//while we are busted, we won't draw power
+	update_icon()
+
+/obj/machinery/food_replicator/proc/set_fix()
+	stat &= ~BROKEN
+	use_power = 1	//let the power flow through you!
+	update_icon()
+
+/obj/machinery/food_replicator/emp_act(severity)
+	..()
+	switch(severity)
+		if(1)
+			if(prob(50))	//50% chance to wipe a random memory slot we have access to
+				clear_memory(rand(1, memory_slots), 1)
+			//break the machine so they have to repair it with cables
+			set_broken()
+		if(2)
+			if(prob(50))	//50% chance to wipe all memory slots, otherwise it wipes a random one we have access to
+				clear_memory(-1)
+			else
+				clear_memory(rand(1,memory_slots), 1)
+			//break the machine so they have to repair it with cables
+			set_broken()
+
+/obj/machinery/food_replicator/power_change()
+	..()
+	update_icon()
+
+/obj/machinery/food_replicator/update_icon()
+	if(stat & BROKEN)
+		icon_state = broken_icon
+	else if(panel_open)		//this is so open panels won't magically appear to close just because the power changed
+		icon_state = open_icon
+	else
+		if(stat & NOPOWER)
+			spawn(rand(0, 15))
+				icon_state = off_icon
+		else
+			icon_state = on_icon
+
+/obj/machinery/food_replicator/proc/eject_dish()
+	if(!loaded)
+		return
+	loaded.forceMove(get_turf(src))
+	loaded = null
+
+/obj/machinery/food_replicator/ui_interact(mob/user, ui_key = "main", datum/nanoui/ui = null, force_open = 1)
+	user.set_machine(src)
+
+	ui = nanomanager.try_update_ui(user, src, ui_key, ui, force_open)
+	if(!ui)
+		ui = new(user, src, ui_key, "food_replicator.tmpl", name, 400, 500)
+		ui.open()
+
+/obj/machinery/food_replicator/ui_data(mob/user, ui_key = "main", datum/topic_state/state = default_state)
+	var/data[0]
+
+	data["points"] = food_paste
+	data["loaded"] = null
+	data["memory"] = null
+	data["slots"] = memory_slots
+
+	if(loaded)
+		var/recycle_value = scan_food(1) * recycle_rate
+		var/rep_cost = scan_food()
+
+		var/list/loaded_data = list()
+		loaded_data["name"] = loaded.name
+		loaded_data["duped"] = loaded.duped
+		loaded_data["recycle_value"] = recycle_value
+		loaded_data["rep_cost"] = rep_cost
+
+		data["loaded"] = loaded_data
+
+	var/list/memory_data = list()
+	memory_data.Add(list(list("slot" = 1, "item_name" = memory1["name"], "cost" = get_cost_from_memory(memory1["reagents"]))))
+	memory_data.Add(list(list("slot" = 2, "item_name" = memory2["name"], "cost" = get_cost_from_memory(memory2["reagents"]))))
+	memory_data.Add(list(list("slot" = 3, "item_name" = memory3["name"], "cost" = get_cost_from_memory(memory3["reagents"]))))
+	memory_data.Add(list(list("slot" = 4, "item_name" = memory4["name"], "cost" = get_cost_from_memory(memory4["reagents"]))))
+	data["memory"] = memory_data
```

Comment:
Aside from being copy-paste-y, this is a very... circuitous way to assemble what should be a very simple list of lists. It *could* (but shouldn't, because you'd still be copy-pasting) look like this:
```
data["memory"] = list(
  list("slot" = 1, "etcetera" = 1),
  list("slot" = 2, "etcetera" = 2)
)
```
... but don't make it look like that, because you want to be iterating over an arbitrarily-sized list, rather than 4 fixed vars.